funsor/optimizer.py

```diff
@@ -1,197 +1,100 @@
-        if isinstance(term, Reduce):
-            if not reduce_op or not (reduce_op, op) in DISTRIBUTIVE_OPS:
-                reduce_op = term.op
-            if term.op == reduce_op and (reduce_op, op) in DISTRIBUTIVE_OPS:
-                reduce_terms.append(term)
-                reduced_vars = reduced_vars | term.reduced_vars
-            else:
-                remaining_terms.append(term)
-        else:
-            remaining_terms.append(term)
-
-    if len(reduce_terms) > 1:
-        new_finitary_term = Finitary(op, tuple(term.arg for term in reduce_terms))
-        remaining_terms.append(Reduce(reduce_op, new_finitary_term, reduced_vars))
-        return Finitary(op, tuple(remaining_terms))
+        if red_op in (v.red_op, nullop) and (v.red_op, bin_op) in DISTRIBUTIVE_OPS:
+            new_terms = terms[:i] + (Contraction(v.red_op, v.bin_op, frozenset(), *v.terms),) + terms[i+1:]
+            return Contraction(v.red_op, bin_op, v.reduced_vars, *new_terms).reduce(red_op, reduced_vars)
+
+        if v.red_op in (red_op, nullop) and bin_op in (v.bin_op, nullop):
+            red_op = v.red_op if red_op is nullop else red_op
+            bin_op = v.bin_op if bin_op is nullop else bin_op
+            new_terms = terms[:i] + v.terms + terms[i+1:]
+            return Contraction(red_op, bin_op, reduced_vars | v.reduced_vars, *new_terms)
 
     return None
 
 
-@dispatched_interpretation
+unfold.register(Contraction, AssociativeOp, AssociativeOp, frozenset, Variadic[Funsor])(
+    lambda r, b, v, *ts: unfold(Contraction, r, b, v, tuple(ts)))
+
+
+@interpreter.dispatched_interpretation
 def optimize(cls, *args):
     result = optimize.dispatch(cls, *args)
     if result is None:
-        result = lazy(cls, *args)
+        result = eager(cls, *args)
     return result
 
 
 # TODO set a better value for this
 REAL_SIZE = 3  # the "size" of a real-valued dimension passed to the path optimizer
 
 
-@optimize.register(Reduce, AssociativeOp, Funsor, frozenset)
-def optimize_reduction_trivial(op, arg, reduced_vars):
-    if not reduced_vars:
-        return arg
-    return None
+optimize.register(Contraction, AssociativeOp, AssociativeOp, frozenset, Variadic[Funsor])(
+    lambda r, b, v, *ts: optimize(Contraction, r, b, v, tuple(ts)))
 
 
-@optimize.register(Reduce, AssociativeOp, Binary, frozenset)
-@eager.register(Reduce, AssociativeOp, Binary, frozenset)
-def optimize_reduce_binary_exp(op, arg, reduced_vars):
-    if op is not ops.add or arg.op is not ops.mul or \
-            not isinstance(arg.lhs, Unary) or arg.lhs.op is not ops.exp:
-        return None
-    return Integrate(arg.lhs.arg, arg.rhs, reduced_vars)
+@optimize.register(Contraction, AssociativeOp, AssociativeOp, frozenset, Funsor, Funsor)
+@optimize.register(Contraction, AssociativeOp, AssociativeOp, frozenset, Funsor)
+def eager_contract_base(red_op, bin_op, reduced_vars, *terms):
+    return None  # eager.dispatch(Contraction, red_op, bin_op, reduced_vars, *terms)
```

Comment:
what does this comment mean?

Reply:
It's just dead code from the switch from `lazy` to `eager` in `optimize`

Reply:
I'll remove